File: backend/garden.py
"""
The garden (POPPY_RETENTION_ENGINE §3.1).

The long game: the thing that makes leaving painful after two months. Every
meaningful call grows something, and the garden is the rendered form of the
user's history with her.

**It carries no number.** Not one. §3's split is that the garden holds the
meaning and the daily layer (bloom.py) holds all the counting, and the moment a
relationship gets a score attached the emotional register breaks. There is
deliberately no total, no percentage and no score anywhere in this module's
output.

Why a garden rather than XP: it is sunk cost that feels like a possession rather
than a score. You don't quit a game to lose points; you quit a garden to abandon
something you grew. Same mechanism, far stronger valence, and it doesn't cheapen
the relationship.

**Nothing here ever wilts.** Miss a week and nothing dies, it just doesn't grow.
A wilting garden is streak-shame with a paint job: it produces a spike and then a
cliff, and it is the one thing §3.1 marks red about this section.
"""

import uuid
from datetime import datetime

import companion
import streak

# A call plants a bud; a call with a real moment in it blooms (§3.1).
BUD = "bud"
BLOOM = "bloom"

# Flowers have identity, so the garden reads as an emotional record of the year
# rather than a scoreboard. The kind comes from the mood mode the call entered
# with; a vent-call flower should not look like a hype-call flower.
KINDS = {
    "vent":  {"label": "vent",      "petals": 5, "hue": "#c2456b"},
    "hype":  {"label": "hype",      "petals": 8, "hue": "#f2913c"},
    "wind":  {"label": "wind down", "petals": 6, "hue": "#5aa9d6"},
    "plan":  {"label": "plan",      "petals": 4, "hue": "#63a85b"},
    "talk":  {"label": "talk",      "petals": 6, "hue": "#e0554f"},
    "ritual": {"label": "ritual",   "petals": 7, "hue": "#8b7bd8"},
}
DEFAULT_KIND = "talk"

# Seasons shift the garden with the real calendar (§3.1): temporal landmarks made
# visual. Northern-hemisphere months; the label is what the UI tints from.
_SEASONS = (
    ((12, 1, 2), "winter"),
    ((3, 4, 5), "spring"),
    ((6, 7, 8), "summer"),
    ((9, 10, 11), "autumn"),
)

_MAX_FLOWERS = 800  # a couple of years of daily calls


def season_for(when: datetime | None = None) -> str:
    month = (when or datetime.now()).month
    for months, name in _SEASONS:
        if month in months:
            return name
    return "summer"


def _flowers() -> list[dict]:
    return list(companion.profile().get("garden") or [])


def plant(kind: str | None = None, bloomed: bool = False, when: datetime | None = None) -> dict:
    """Grow one flower for a call.

    A bud for turning up, a bloom for a call that had something real in it. The
    two are stored the same way so a bud is never a failure, just a smaller thing
    that happened.
    """
    when = when or datetime.now()
    kind = kind if kind in KINDS else DEFAULT_KIND
    flower = {
        "id": uuid.uuid4().hex[:10],
        "kind": kind,
        "state": BLOOM if bloomed else BUD,
        "date": streak.today(when).isoformat(),
        "season": season_for(when),
        # Stable per-flower jitter so the field looks grown rather than plotted,
        # and looks the same every time it is drawn.
        "seed": int(uuid.uuid4().int % 10_000),
    }
    flowers = _flowers() + [flower]
    companion.update(garden=flowers[-_MAX_FLOWERS:])
    return flower


def bloom_last(when: datetime | None = None) -> dict | None:
    """Open today's bud, if a real moment landed after it was planted."""
    flowers = _flowers()
    day = streak.today(when or datetime.now()).isoformat()
    for f in reversed(flowers):
        if f.get("date") == day and f.get("state") == BUD:
            f["state"] = BLOOM
            companion.update(garden=flowers)
            return f
    return None


def arrange(positions: dict) -> int:
    """Save where the user put their flowers. Returns how many moved.

    Until now the garden arranged itself and the user only watched it. Placing
    things yourself is the one motivation this product had no answer for: it is
    why people rearrange a room or keep a shelf a particular way, and unlike a
    level cap it never runs out, because the novelty comes from them rather than
    from us.

    `x` and `y` are 0-1 in field space, not pixels, so an arrangement survives a
    resized window and a different screen.
    """
    flowers = _flowers()
    by_id = {f.get("id"): f for f in flowers}
    moved = 0
    for fid, pos in (positions or {}).items():
        f = by_id.get(fid)
        if not f or not isinstance(pos, dict):
            continue
        try:
            x, y = float(pos["x"]), float(pos["y"])
        except (KeyError, TypeError, ValueError):
            continue
        # Clamped rather than rejected: a flower dragged off the edge should end
        # up at the edge, not vanish.
        f["x"] = min(max(x, 0.0), 1.0)
        f["y"] = min(max(y, 0.0), 1.0)
        moved += 1
    if moved:
        companion.update(garden=flowers)
    return moved


MAX_LABEL = 60


def label(flower_id: str, text: str) -> dict | None:
    """Name one flower. "the day I quit."

    The garden already records *that* something happened; a label is the user
    saying what it meant. That is theirs to write, not ours to infer, and it is
    the second thing in the product they make rather than receive.

    An empty label clears it, so a name is never permanent by accident.
    """
    flowers = _flowers()
    text = (text or "").strip()[:MAX_LABEL]
    for f in flowers:
        if f.get("id") != flower_id:
            continue
        if text:
            f["label"] = text
        else:
            f.pop("label", None)
        companion.update(garden=flowers)
        return f
    return None


def labelled_count() -> int:
    """How many flowers carry a name. A count only: the labels themselves are the
    user's own words about their own life and never reach the analytics log."""
    return sum(1 for f in _flowers() if f.get("label"))


def state(limit: int = 400) -> dict:
    """What the renderer needs.

    Note what is absent: no count, no total, no level, no percentage. §3 keeps
    every number on the other surface, and the moment one appears here the user
    starts gardening for the score instead of talking.
    """
    flowers = _flowers()[-limit:]
    return {
        "flowers": flowers,
        "season": season_for(),
        "kinds": KINDS,
        # True only before anything has ever grown, so the UI can stay hidden
        # rather than render an empty plot (§8: first sight is never zero).
        "empty": not flowers,
    }


def year_in_review(when: datetime | None = None) -> dict:
    """"My year with Poppy" (§3.1): the private-by-default share artifact.

    Returns the shape of the year, never a scoreboard: which kinds of call the
    year was made of, and across which seasons.
    """
    when = when or datetime.now()
    year = str(when.year)
    flowers = [f for f in _flowers() if str(f.get("date", ""))[:4] == year]
    by_kind: dict[str, int] = {}
    by_season: dict[str, int] = {}
    for f in flowers:
        by_kind[f["kind"]] = by_kind.get(f["kind"], 0) + 1
        by_season[f["season"]] = by_season.get(f["season"], 0) + 1
    return {
        "year": year,
        "flowers": flowers,
        "by_kind": by_kind,
        "by_season": by_season,
        "empty": not flowers,
    }

File: mobile/src/core/garden.ts
/**
 * The garden — the port of backend/garden.py.
 *
 * One flower per call: a bud for turning up, a bloom for a call that had something
 * real in it. Both stored the same way, so a bud is never a failure, just a smaller
 * thing that happened.
 *
 * Note what this deliberately does not return: no count, no total, no level, no
 * percentage. Every number lives on the other surface. The moment one appears here
 * the user starts gardening for a score instead of talking.
 */

import * as companion from './companion';
import { streakDay } from './streak';

export const BUD = 'bud';
export const BLOOM = 'bloom';

export const KINDS: Record<
  string,
  { label: string; petals: number; hue: string; rare?: boolean }
> = {
  vent: { label: 'vent', petals: 5, hue: '#c2456b' },
  hype: { label: 'hype', petals: 8, hue: '#f2913c' },
  wind: { label: 'wind down', petals: 6, hue: '#5aa9d6' },
  plan: { label: 'plan', petals: 4, hue: '#63a85b' },
  talk: { label: 'talk', petals: 6, hue: '#e0554f' },
  ritual: { label: 'ritual', petals: 7, hue: '#8b7bd8' },
  // The Long Year. Obtainable exactly one way, by being here a year, which is what
  // makes it worth anything. Never planted by an ordinary call.
  longyear: { label: 'the long year', petals: 12, hue: '#e8b53f', rare: true },
};

export const DEFAULT_KIND = 'talk';
const MAX_FLOWERS = 800;

export type Flower = {
  id: string;
  kind: string;
  state: string;
  date: string;
  season: string;
  seed: number;
  x?: number;
  y?: number;
  label?: string;
};

/** Seasons shift the garden with the real calendar: landmarks made visual. */
export function seasonFor(when = new Date()): string {
  const m = when.getMonth() + 1;
  if ([12, 1, 2].includes(m)) return 'winter';
  if ([3, 4, 5].includes(m)) return 'spring';
  if ([6, 7, 8].includes(m)) return 'summer';
  return 'autumn';
}

async function flowers(): Promise<Flower[]> {
  return ((await companion.profile()).garden ?? []) as unknown as Flower[];
}

function newId(): string {
  return `${Date.now().toString(36)}${Math.random().toString(36).slice(2, 6)}`;
}

export async function plant(
  kind: string | null = null,
  bloomed = false,
  when = new Date(),
): Promise<Flower> {
  const k = kind && KINDS[kind] ? kind : DEFAULT_KIND;
  const flower: Flower = {
    id: newId(),
    kind: k,
    state: bloomed ? BLOOM : BUD,
    date: streakDay(when),
    season: seasonFor(when),
    // Stable per-flower jitter so the field looks grown rather than plotted, and
    // looks the same every time it is drawn.
    seed: Math.floor(Math.random() * 10000),
  };
  const all = [...(await flowers()), flower];
  await companion.update({ garden: all.slice(-MAX_FLOWERS) as unknown as Record<string, unknown>[] });
  return flower;
}

/** Turn today's bud into a bloom, if there is one. */
export async function bloomLast(when = new Date()): Promise<Flower | null> {
  const all = await flowers();
  const today = streakDay(when);
  for (let i = all.length - 1; i >= 0; i--) {
    if (all[i].date === today) {
      all[i].state = BLOOM;
      await companion.update({ garden: all as unknown as Record<string, unknown>[] });
      return all[i];
    }
  }
  return null;
}

/** The rare one. Planted once, and only by having been here a year. */
export async function plantLongYear(when = new Date()): Promise<Flower | null> {
  const p = await companion.profile();
  if (p.long_year_marked) return null;
  const flower = await plant('longyear', true, when);
  await companion.update({ long_year_marked: true });
  return flower;
}

/** Drag-to-arrange: {id: {x, y}} in normalised 0..1 coordinates. */
export async function arrange(
  positions: Record<string, { x: number; y: number }>,
): Promise<number> {
  const all = await flowers();
  let moved = 0;
  for (const f of all) {
    const pos = positions[f.id];
    if (!pos) continue;
    const x = Number(pos.x);
    const y = Number(pos.y);
    if (Number.isNaN(x) || Number.isNaN(y)) continue;
    f.x = Math.max(0, Math.min(1, x));
    f.y = Math.max(0, Math.min(1, y));
    moved++;
  }
  if (moved) await companion.update({ garden: all as unknown as Record<string, unknown>[] });
  return moved;
}

/** Naming a day: the user's own words on their own history. */
export async function label(id: string, text: string): Promise<Flower | null> {
  const all = await flowers();
  const f = all.find((x) => x.id === id);
  if (!f) return null;
  const clean = (text || '').trim().slice(0, 60);
  if (clean) f.label = clean;
  else delete f.label;
  await companion.update({ garden: all as unknown as Record<string, unknown>[] });
  return f;
}

export async function labelledCount(): Promise<number> {
  return (await flowers()).filter((f) => f.label).length;
}

export async function state(limit = 400): Promise<Record<string, unknown>> {
  const all = (await flowers()).slice(-limit);
  return {
    flowers: all,
    season: seasonFor(),
    kinds: KINDS,
    // True only before anything has ever grown, so the UI can stay hidden rather
    // than render an empty plot: first sight is never zero.
    empty: all.length === 0,
  };
}

export async function yearInReview(when = new Date()): Promise<Record<string, unknown>> {
  const all = await flowers();
  const year = String(when.getFullYear());
  const mine = all.filter((f) => f.date.startsWith(year));
  const byKind: Record<string, number> = {};
  for (const f of mine) byKind[f.kind] = (byKind[f.kind] ?? 0) + 1;
  return {
    year,
    total: mine.length,
    blooms: mine.filter((f) => f.state === BLOOM).length,
    by_kind: byKind,
    labelled: mine.filter((f) => f.label).length,
    rare: mine.filter((f) => KINDS[f.kind]?.rare).length,
  };
}
